### DISCUSSION TITLE
construct a building

### DISCUSSION THREAD

**Discussion by zewa666** [Annoraaq/grid-engine#240] [Q&A]
hello, I'd like to add building construction to my game. So my general idea was to:

* open a dialog and let user pick a building
* draw the tiles involved (similar to placing them in tiled) on the players facing position with a lower alpha
* take over controls from player and move the blueprint instead
* upon ok input place the actual building

so far the idea, it should resemble what kinda classic rts do.
now tbh I have absolutely no idea about how to construct the template out of multiple tiles. would you create a tilemap for each individual building? keep in mind different layers (ground, top) are involved as well.

@Annoraaq did you do anything similar already for your games?

**Comment by zewa666** [comment]
perhaps to make it a bit more clearer here a few screenshots:

This is the placement mode:
![image](https://user-images.githubusercontent.com/2861414/157434760-ea1a3ba7-b18a-475e-a6c8-aab3575d4ded.png)


This is what I envision it after it's placed:
![2022-03-09_12h35_13](https://user-images.githubusercontent.com/2861414/157434368-2b83e25e-6ab8-4447-97da-5d7a3629a63a.png)

And here the different layers of how I'd build it in Tiled as an example:
![2022-03-09_12h36_08](https://user-images.githubusercontent.com/2861414/157434848-5e68bcdd-60a1-401c-9e43-d98d2f667253.png)
![2022-03-09_12h36_16](https://user-images.githubusercontent.com/2861414/157434854-b3d2b168-2d9a-4f48-928b-787609d7da88.png)
![2022-03-09_12h36_20](https://user-images.githubusercontent.com/2861414/157434860-085a863b-8ed0-48fd-bd8d-d1ccd606d50d.png)

**Comment by Annoraaq** [comment]
I thought about that, too. But I never actually implemented it, so I can only give you some high level ideas I have about this. I thought about this to manually create templates that could be used by a procedural dungeon/map generator.

My idea was to create a tile map in Tiled for each template and then load them in phaser and merge it with your existing tile map on demand (when the user builds something). I don't know if there are any convenient methods in phaser or if you have to do it manually. But you can definitely manipulate the tile maps in memory with Phaser.

One thing to also check would be the size of the templates. To my knowledge Phaser only supports inline tile sets. Meaning that the same tile set is included in every template. But there is probably a way around this.

**Reply by zewa666** [reply]
just an update, I didnt really come up with a satisfactory solution so instead I've changed the game design so that there are fixed locations for the buildings.

now I can build them right away and setVisible(false) the tiles and objects (doors, lanterns) at the begin and later on change visibility after building succeeded.
